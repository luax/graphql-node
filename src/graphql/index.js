const merge = require("lodash.merge");
const { AuthenticationError } = require("./errors");
const Query = require("./Query");
const Book = require("./Book");
const User = require("./User");
const Author = require("./Author");

const typeDefs = [
  Query.typeDefs,
  Author.typeDefs,
  Book.typeDefs,
];

const resolvers = merge(
  Query.resolvers,
  Author.resolvers,
  Book.resolvers,
);

const context = async ({ req }) => {
  const user = User.getUser(req);
  if (!user)
    throw new AuthenticationError("you must be logged in to query this schema");
  return {
    // Create new loaders per request
    loaders: merge(Book.loaders(user), Author.loaders(user)),
  };
};

module.exports = {
  typeDefs,
  resolvers,
  context,
};
