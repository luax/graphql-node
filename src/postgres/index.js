const client = require("./client");
const { sql } = require("slonik");

module.exports = {
  client,
  sql,
};
