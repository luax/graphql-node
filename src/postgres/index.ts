export { default as client, QueryResultRowType } from "./client";
export { sql } from "slonik";
