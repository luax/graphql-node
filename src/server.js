const { ApolloServer } = require("apollo-server");
const { typeDefs, resolvers, context } = require("./graphql");
const { formatError } = require("./graphql/errors");

const createServer = ({ typeDefs, resolvers, context }) => {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    context,
    formatError,
    formatResponse: response => {
      return response;
    },
    tracing: process.env.NODE_ENV === "development",
  });
  return server;
};

const startServer = async () => {
  const server = createServer({ typeDefs, resolvers, context });
  const httpServer = await server.listen({
    port: process.env.PORT,
  });
  console.log(
    `ðŸš€  Server ready at ${httpServer.url} (env "${process.env.NODE_ENV}")`,
  );
  return httpServer;
};

module.exports = {
  createServer,
  startServer,
};
