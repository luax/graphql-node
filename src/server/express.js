const express = require("express");
const compression = require("compression");
const cors = require("cors");
const helmet = require("helmet");

const getUrl = ({ address, port }) => {
  let host = address;
  if (address === "" || address === "::") host = "localhost";
  return `http://${host}:${port}/`;
};

const setupExpress = () => {
  const app = express();
  app.use(compression());
  app.use(helmet());
  app.use(cors());
  app.get("/", (_req, res) => {
    res.send("Hello World!");
  });
  return app;
};

const startServer = async apolloServer => {
  const app = setupExpress();
  apolloServer.applyMiddleware({ app });
  const httpServer = await app.listen({
    port: process.env.PORT,
  });
  const url = getUrl(httpServer.address());
  console.log(`ðŸš€  Server ready at ${url} (env "${process.env.NODE_ENV}")`);
  console.log(`Health check at: ${url}/.well-known/apollo/server-health`);
  return { url, server: httpServer, apolloServer, app };
};

module.exports = {
  startServer,
};
