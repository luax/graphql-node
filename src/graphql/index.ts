import merge from "lodash/merge";
import context from "./context";
import Root from "./Root";
import User, { UserLoaders } from "./User";
import Book, { BookLoaders } from "./Book";
import Author, { AuthorLoaders } from "./Author";
import { IResolvers } from "apollo-server";
import { DocumentNode } from "graphql";
import { Model } from "./types";

const typeDefs: DocumentNode[] = [
  Root.typeDefs,
  User.typeDefs,
  Author.typeDefs,
  Book.typeDefs,
];

const resolvers: IResolvers = merge(
  Root.resolvers,
  User.resolvers,
  Author.resolvers,
  Book.resolvers,
);

export type Loaders = BookLoaders & UserLoaders & AuthorLoaders;

const models: Model<BookLoaders | UserLoaders | AuthorLoaders>[] = [Author];

const loaders = (): Loaders =>
  merge(Book.loaders(), User.loaders(), Author.loaders());

export default {
  typeDefs,
  resolvers,
  context: context(loaders),
  loaders,
  models,
};
