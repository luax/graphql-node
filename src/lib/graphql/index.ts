import graphqlFields from "graphql-fields";
import { GraphQLResolveInfo } from "graphql";

const selectFields = (
  includedFields: Set<string>,
  selectableFields: Set<string>,
): ((info: GraphQLResolveInfo) => string[]) => {
  const allFields = [...includedFields, ...selectableFields].sort();
  return (info: GraphQLResolveInfo): string[] => {
    if (!info) {
      return allFields;
    }
    const fields = Object.keys(graphqlFields(info));
    const selectedFields = fields.filter(field => selectableFields.has(field));
    const sortedFields = [...includedFields, ...selectedFields].sort();
    return sortedFields;
  };
};

export { selectFields };
