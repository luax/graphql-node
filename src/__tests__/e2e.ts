import graphql from "../graphql";
import { startServer, createApolloServer } from "../server";
import { client } from "../postgres";
import axios from "axios";
import { ServerInfo } from "../server/express";

describe("e2e", () => {
  let serverInfo: ServerInfo;

  beforeAll(() => {
    serverInfo = startServer(createApolloServer(graphql));
    client.initialize();
  });

  afterAll(async () => {
    serverInfo.server.close();
    await client.end();
  });

  it("gets a list of books", async () => {
    const response = await axios.post(`${serverInfo.url}graphql`, {
      query: `
        {
          books {
            id
            title
            author {
              id
              name
              books {
                id
                title
                author {
                  id
                  name
                }
              }
            }
          }
        }
      `,
    });
    expect(response.data).toMatchSnapshot();
  });

  it("gets a book", async () => {
    const response = await axios.post(`${serverInfo.url}graphql`, {
      query: `
        {
          book(id: 1) {
            id
            title,
            author {
              name
              books {
                id
                title
              }
            }
          }
        }
      `,
    });
    expect(response.data).toMatchSnapshot();
  });

  it("gets an author", async () => {
    const response = await axios.post(`${serverInfo.url}graphql`, {
      query: `
        {
          author(id: 1) {
            id
            name
            books {
              id
              title,
              author {
                id
                name
              }
            }
          }
        }
      `,
    });
    expect(response.data).toMatchSnapshot();
  });
});
