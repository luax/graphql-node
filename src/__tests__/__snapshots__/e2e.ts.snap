// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e gets a book 1`] = `
Object {
  "data": Object {
    "book": Object {
      "author": Object {
        "books": Array [
          Object {
            "id": "1",
            "title": "Harry Potter and the Philosopher’s stonee",
          },
          Object {
            "id": "2",
            "title": "Harry Potter and the Chamber of Secrets",
          },
          Object {
            "id": "3",
            "title": "Harry Potter and the Chamber of Secrets",
          },
          Object {
            "id": "4",
            "title": "Harry Potter and the Prisoner of Azkaban",
          },
          Object {
            "id": "5",
            "title": "Harry Potter and the Goblet of Fire",
          },
          Object {
            "id": "6",
            "title": "Harry Potter and the Order of the Phoenix",
          },
          Object {
            "id": "7",
            "title": "Harry Potter and the Half-Blood Prince",
          },
          Object {
            "id": "9",
            "title": "Harry Potter and the Deathly Hallows",
          },
          Object {
            "id": "10",
            "title": "Fantastic Beasts and Where to Find Them",
          },
          Object {
            "id": "11",
            "title": "Quidditch Through the Ages",
          },
          Object {
            "id": "12",
            "title": "The Tales of Beedle the Bard",
          },
        ],
        "name": "J.K. Rowling",
      },
      "id": "1",
      "title": "Harry Potter and the Philosopher’s stonee",
    },
  },
}
`;

exports[`e2e gets a list of books 1`] = `
Object {
  "data": Object {
    "books": Array [
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "1",
        "title": "Harry Potter and the Philosopher’s stonee",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "2",
        "title": "Harry Potter and the Chamber of Secrets",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "3",
        "title": "Harry Potter and the Chamber of Secrets",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "4",
        "title": "Harry Potter and the Prisoner of Azkaban",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "5",
        "title": "Harry Potter and the Goblet of Fire",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "6",
        "title": "Harry Potter and the Order of the Phoenix",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "7",
        "title": "Harry Potter and the Half-Blood Prince",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "2",
                "name": "Michael Crichton",
              },
              "id": "8",
              "title": "Jurassic Park",
            },
          ],
          "id": "2",
          "name": "Michael Crichton",
        },
        "id": "8",
        "title": "Jurassic Park",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "9",
        "title": "Harry Potter and the Deathly Hallows",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "10",
        "title": "Fantastic Beasts and Where to Find Them",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "11",
        "title": "Quidditch Through the Ages",
      },
      Object {
        "author": Object {
          "books": Array [
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "1",
              "title": "Harry Potter and the Philosopher’s stonee",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "2",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "3",
              "title": "Harry Potter and the Chamber of Secrets",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "4",
              "title": "Harry Potter and the Prisoner of Azkaban",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "5",
              "title": "Harry Potter and the Goblet of Fire",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "6",
              "title": "Harry Potter and the Order of the Phoenix",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "7",
              "title": "Harry Potter and the Half-Blood Prince",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "9",
              "title": "Harry Potter and the Deathly Hallows",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "10",
              "title": "Fantastic Beasts and Where to Find Them",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "11",
              "title": "Quidditch Through the Ages",
            },
            Object {
              "author": Object {
                "id": "1",
                "name": "J.K. Rowling",
              },
              "id": "12",
              "title": "The Tales of Beedle the Bard",
            },
          ],
          "id": "1",
          "name": "J.K. Rowling",
        },
        "id": "12",
        "title": "The Tales of Beedle the Bard",
      },
      Object {
        "author": null,
        "id": "13",
        "title": "Ghost Author",
      },
    ],
  },
}
`;

exports[`e2e gets an author 1`] = `
Object {
  "data": Object {
    "author": Object {
      "books": Array [
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "1",
          "title": "Harry Potter and the Philosopher’s stonee",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "2",
          "title": "Harry Potter and the Chamber of Secrets",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "3",
          "title": "Harry Potter and the Chamber of Secrets",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "4",
          "title": "Harry Potter and the Prisoner of Azkaban",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "5",
          "title": "Harry Potter and the Goblet of Fire",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "6",
          "title": "Harry Potter and the Order of the Phoenix",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "7",
          "title": "Harry Potter and the Half-Blood Prince",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "9",
          "title": "Harry Potter and the Deathly Hallows",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "10",
          "title": "Fantastic Beasts and Where to Find Them",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "11",
          "title": "Quidditch Through the Ages",
        },
        Object {
          "author": Object {
            "id": "1",
            "name": "J.K. Rowling",
          },
          "id": "12",
          "title": "The Tales of Beedle the Bard",
        },
      ],
      "id": "1",
      "name": "J.K. Rowling",
    },
  },
}
`;
